include/CL/sycl/detail/buffer.hpp

```diff
@@ -130,6 +130,22 @@ class buffer_impl
               void* host_ptr,
               bool is_svm_ptr = false);
 
+  buffer_impl(
+    void* _buffer_pointer_source,
+    void* _host_memory_source,
```

Comment:
This is not really a good approach:

In general, you would expect a SYCL object to manage memory and bind memory lifetime to object lifetime using reference counting semantics.

With this implementation, when the last copy of the original buffer goes out of scope, the subbuffer will suddenly point to invalid memory.
When the subbuffer goes out of scope no cleanup will ever happen because memory is still managed by the original buffer.

You need to use the same `buffer_impl` as the orignal buffer.

Reply:
(This is also important for data dependency calculation, as this is done per `buffer_impl` object)

Reply:
I assumed that the contract was that the subbuffers are subservient to the parent buffer, so that when the parent buffer dies all the subbuffers do too. 

I suppose the subbuffers may live longer the parents if one passes their reference out of the scope.

Reply:
I don't completely understand what you are saying, let me think about it and understand the current implementation.

Reply:
In the new implementation, will only the memory region owned by the subbuffer continue to live after the parent dies, or the entire buffer?

Reply:
Since we cannot partially deallocate an allocated memory block on all APIs that I am aware of, allocation and deallocation will always have to be done for the entire buffer.